target/linux/realtek/files-5.10/drivers/net/phy/rtl83xx-phy.c

```diff
@@ -2952,13 +3107,594 @@ int rtl9300_sds_cmu_band_get(int sds)
+	pr_info("%s configuring RTL8226 on port %d\n", __func__, phy_addr);
+	if (dev->of_node) {
+		dn = dev->of_node;
+
+		if (of_property_read_u32(dn, "sds", &sds_num))
+			return 0;  /* Not the base address */
+	} else {
+		dev_err(dev, "No DT node.\n");
+		return -EINVAL;
+	}
+
+	pr_info("%s: port %d, SerDes is %d\n", __func__, phy_addr, sds_num);
+	pr_info("%s CMU BAND is %d\n", __func__, rtl9300_sds_cmu_band_get(sds_num));
+
+	saved_state  = disable_polling(phy_addr);
+
+	/* Disable MAC */
+	sw_w32_mask(0, 1, RTL930X_MAC_FORCE_MODE_CTRL + 4 * phy_addr);
+	mdelay(20);
+
+	/*
+	 * On the RTL8226 no need to change the polarity, see
+	 * dal_longan_construct_macConfig_init
+	 */
+
+	/* Turn Off Serdes */
+	rtl9300_force_sds_mode(sds_num, PHY_INTERFACE_MODE_NA);
+
+	pr_info("%s PATCHING SerDes %d\n", __func__, sds_num);
+
+	if (sds_num % 2) {
+		rtl9300_sds_patch(sds_num, rtl9300_a_sds_10g_hisgmii_lane1,
+				  sizeof(rtl9300_a_sds_10g_hisgmii_lane1) / sizeof(sds_config));
+	} else {
+		rtl9300_sds_patch(sds_num, rtl9300_a_sds_10g_hisgmii_lane0,
+				  sizeof(rtl9300_a_sds_10g_hisgmii_lane0) / sizeof(sds_config));
+	}
+
+	/* Configure PHY from phy_construct_config_init -> rtl8226_config */
+	rtl9300_rtl8226_phy_setup(phydev);
+
+	/* Configure link to MAC */
+	rtl9300_serdes_mac_link_config(sds_num, true, true); /* MAC Construct */
+
+	/* Re-Enable MAC */
+	sw_w32_mask(1, 0, RTL930X_MAC_FORCE_MODE_CTRL + 4 * phy_addr);
+
+	/* Set initial RX calibration parameter, but do not perform actual calibration */
+	rtl9300_do_rx_calibration_1(sds_num, phy_mode);
+
+	/* Re-enable SDS with new mode */
+	rtl9300_force_sds_mode(sds_num, phy_mode);
+
+	rtl9300_sds_tx_config(sds_num, phy_mode);
+
+	/* Re-enable polling */
+	resume_polling(saved_state);
+
+	return 0;
+}
+
+int rtl9300_rtl8226_mode_set(int port, int sds_num, phy_interface_t phy_mode)
+{
+	pr_info("%s setting serdes %d to mode %s +++++\n", __func__, sds_num, phy_modes(phy_mode));
+	/* Disable MAC */
+	sw_w32_mask(0, 1, RTL930X_MAC_FORCE_MODE_CTRL + 4 * port);
+	mdelay(20);
+
+	/* Turn Off Serdes */
+	rtl9300_force_sds_mode(sds_num, PHY_INTERFACE_MODE_NA);
+
+	/* Configure link to MAC */
+	rtl9300_serdes_mac_link_config(sds_num, true, true); /* MAC Construct */
+
+	/* Re-Enable MAC */
+	sw_w32_mask(1, 0, RTL930X_MAC_FORCE_MODE_CTRL + 4 * port);
+
+	/* Set initial RX calibration parameter, but do not perform actual calibration */
+	rtl9300_do_rx_calibration_1(sds_num, PHY_INTERFACE_MODE_HSGMII);
+
+	/* Re-enable SDS with new mode */
+	rtl9300_force_sds_mode(sds_num, phy_mode);
+
+	rtl9300_sds_tx_config(sds_num, PHY_INTERFACE_MODE_HSGMII);
+
+	return 0;
+}
+
+int rtl9300_configure_serdes(int port, int sds_num, phy_interface_t phy_mode)
 {
 	struct device *dev = &phydev->mdio.dev;
 	int phy_addr = phydev->mdio.addr;
 	struct device_node *dn;
 	u32 sds_num = 0;
-	int sds_mode, calib_tries = 0, phy_mode = PHY_INTERFACE_MODE_10GBASER, i;
+	int sds_mode, calib_tries = 0, phy_mode = PHY_INTERFACE_MODE_10GBASER;
```

Comment:
I get `error: 'phy_mode' redeclared as different kind of symbol`

`note: previous definition of 'phy_mode' with type 'phy_interface_t'`